artemis-commons/src/main/java/org/apache/activemq/artemis/utils/SizeAwareMetric.java

```diff
@@ -0,0 +1,240 @@
+      if (lowerMarkElements > maxElements) {
+         throw new IllegalArgumentException("lowerMarkElements must be <= maxElements");
+      }
+      this.maxElements = maxElements;
+      this.lowerMarkElements = lowerMarkElements;
+      this.maxSize = maxSize;
+      this.lowerMarkSize = lowerMarkSize;
+      this.sizeEnabled = maxSize >= 0;
+      this.elementsEnabled = maxElements >= 0;
+   }
+
+   public boolean isOver() {
+      return flag != FREE && flag != NOT_USED;
+   }
+
+   public boolean isOverSize() {
+      return flag == OVER_SIZE;
+   }
+
+   public boolean isOverElements() {
+      return flag == OVER_ELEMENTS;
+   }
+
+   public long getSize() {
+      return size;
+   }
+
+   public boolean isElementsEnabled() {
+      return elementsEnabled;
+   }
+
+   public SizeAwareMetric setElementsEnabled(boolean elementsEnabled) {
+      this.elementsEnabled = elementsEnabled;
+      return this;
+   }
+
+   public long getElements() {
+      return elements;
+   }
+
+   public boolean isSizeEnabled() {
+      return sizeEnabled;
+   }
+
+   public SizeAwareMetric setSizeEnabled(boolean sizeEnabled) {
+      this.sizeEnabled = sizeEnabled;
+      return this;
+   }
+
+   public void setOnSizeCallback(IntConsumer onSize) {
+      this.onSizeCallback = onSize;
+   }
+
+   public void setOnSizeOnlyCallback(IntConsumer onSize) {
+      this.onSizeOnlyCallback = onSize;
+   }
+
+   public void setOverCallback(Runnable over) {
+      this.overCallback = over;
+   }
+
+   public void setParentMetric(SizeAwareMetric parent) {
+      parentMetric = parent;
+   }
+
+   public void setUnderCallback(Runnable under) {
+      this.underCallback = under;
+   }
+
+   protected void over() {
+      if (overCallback != null) {
+         try {
+            overCallback.run();
+         } catch (Throwable e) {
+            logger.warn(e.getMessage(), e);
+         }
+      }
+   }
+
+   protected void under() {
+      if (underCallback != null) {
+         try {
+            underCallback.run();
+         } catch (Throwable e) {
+            logger.warn(e.getMessage(), e);
+         }
+      }
+   }
+
+   private boolean changeFlag(int expected, int newValue) {
+      return flagUpdater.compareAndSet(this, expected, newValue);
+   }
+
+   public long addSize(int delta) {
+      return addSize(delta, false);
+   }
+
+   public long addSizeOnly(int delta) {
+      return addSize(delta, true);
+   }
+
+   public long addSize(int delta, boolean sizeOnly) {
+
+      if (delta == 0) {
+         if (logger.isDebugEnabled()) {
+            logger.debug("SizeAwareMetric ignored with size 0", new Exception("trace"));
+         }
+         return sizeUpdater.get(this);
+      }
+
+      flagUpdater.compareAndSet(this, NOT_USED, FREE);
+
+      if (parentMetric != null) {
+         if (parentMetric != null) {
+            parentMetric.addSize(delta, sizeOnly);
+         }
+      }
+
+      IntConsumer theFunction = sizeOnly ? onSizeOnlyCallback : onSizeCallback;
```

Comment:
Given these only looks to be used to call an equivalent method on anther object, could they be combined into one callback and simply pass the boolean as well? Would need to use a custom callback interface though (or a BiConsumer with boxing). Not really a hardship given it would probably just be supplied with a lambda these days and never actually implemented.

Reply:
yes.. it can be done.. 

I wanted to avoid the Boxing on a BiConsumer, and I forgot for an instant I didn't think about the own interface.. thanks. nice idea.